keps/sig-network/20190415-Autopath API for clusterDNS.md

```diff
@@ -0,0 +1,104 @@
+
+## Motivation
+DNS Search Path expansion on pods using ClusterFirst DNS mode can lead to DNS latency issues and race conditions due to several parallel (The queries are sent in parallel in musl and glibc 2.16+ implementations.) dns queries from the same pod.
+The search path currently includes: 
+
+1. "$NS.svc.$SUFFIX"
+2. "svc.$SUFFIX"
+3. "$SUFFIX"
+4. Host level suffixes, which might be 2 or 3 in number.
+
+Where $NS stands for the namespace that the pod belongs to, $SUFFIX is the Kubernetes cluster suffix.
+
+These search paths are set to make sure: 
+
+ 1. Pods can discover Services in the same namespace using just the service name.
+ 2. Pods can discover Services across namespaces using shorthand of the form "$SVCNAME.$NSNAME"
+ 3. Pods can discover resources within the same cluster.
+
+These searchpaths are included in pods' /etc/resolv.conf by kubelet and are enforced by setting ndots to 5. This means any hostname lookups with fewer than 5 dots will be expanded using all the search paths listed.
+
+When pod issues a query to lookup hostname "service123", it is expanded to 6 queries - one for the original hostname and one with each of the searchpaths appended. Some resolvers issue both A and AAAA queries, so this can be a total of 12 or more queries for every single DNS lookup. When these queries are issued in parallel, they end up at the node with the same source tuple and need to be DNAT'ed increasing the chance of a [netfilter race condition](https://www.weave.works/blog/racy-conntrack-and-dns-lookup-timeouts).
+Even if one of the several queries fails due, the DNS lookup on the client side will fail after a 5s timeout.
+
+### Goals
+Provide a solution to minimize number of DNS queries on the client side, for a DNS lookup. 
+This solution will preserve the current behavior that allows shortname lookups of Kubernetes resources. Make this solution configurable via the API.
+
+### Non-Goals
+Modifying the current Kubernetes DNS Schema is not a goal of this KEP.
+
+
+## Proposal
+
+This proposal introduces a new dnsPolicy "clusterFirstWithAutopath". This can be set by including the line below in podSpec:
+
+ ` dnsPolicy: clusterFirstWithAutopath`
+
+Using this mode will set the searchpath on the pod as `search.$NS.$SUFFIX.k`, where $NS is the namespace of the pod and $SUFFIX is the cluster suffix. k is a one-letter suffix to identify the domain name to be a query that needs search expansion. Since there are [no single-letter TLD so far](http://data.iana.org/TLD/tlds-alpha-by-domain.txt), this suffix will help identify queries needing search expansion. 
```

Comment:
I am thinking over the design here.  Assume that we (one day) want to change the kubernetes DNS schema (we should).  That ends up being as user-facing change.  How does that impact this?

Let's be concrete: Today we have streamlined access to `$S.$N.svc.$Z`, but this has problems.  One proposal was to change to `$S.s.$N.$Z`.  This is strictly ambiguous if there happens to be a namespace `s` and you want to look up namespace `svc`, so we can't really just try both schemas : `foo.s.svc` can mean both. 

So I *think* (please help me be wrong) we need to specify at minimum a schema version in the magic string.  Something like `search.$NS.$SUFFIX.v1.k` so `foo.s.svc` => `foo.s.svc.search.$NS.$SUFFIX.v1.k`, which we can parse as a request for service "foo" in namespace "s".  V2 would be `foo.s.svc.search.$NS.$SUFFIX.v2.k`, which we can (unambiguously!) parse as a request for service "foo" in namespace "svc".

Last thought -- The `.k` TLD seems unlikely to happen, but it's worth thinking about a less terse solutions.  `.kubernetes-autopath-v1` also seems unlikely as a TLD.  Is it better or worse?

Reply:
With node-local cache doing the search expansion, it seemed unnecessary to pass in version if node-local cache version was tied to the schema version. This could still be the case if we moved the logic to CoreDNS and the version would be specified in the CoreFile at cluster bringup time. Like:
`autopath @kubernetes < version>`
The kubernetes plugin will have the logic to construct searchpath based on this version parameter. However, if this is a DNS Server that clusters using different schema versions could talk to, we need the version to be part of the request, so yes, part of the searchpath.

"kubernetes-autopath-v1" is a little long, we want to pick a label that won't result in domain names becoming longer that 255 characters which is the DNSName limit. It is only 10 chars more than "cluster.local". Or we could do "k8s-v1".

Reply:
> The kubernetes plugin will have the logic to construct searchpath based on this version parameter. However, if this is a DNS Server that clusters using different schema versions could talk to, we need the version to be part of the request, so yes, part of the searchpath.

Even apart from this KEP, if clients are talking to the same DNS server expecting different DNS schemas, then that server will already need to know which version to serve up. So, I don't think we need to put it in the magic string - any pod using a different DNS policy would have this same problem, so we'll need to solve that if and when we change the schema version. 

For the label, why not something like "ap.k8s.io" which is relatively short and we already control k8s.io so we can reserve that (and yes, it's three labels not one but I don't think that's an issue).

Reply:
> > The kubernetes plugin will have the logic to construct searchpath based on this version parameter. However, if this is a DNS Server that clusters using different schema versions could talk to, we need the version to be part of the request, so yes, part of the searchpath.
> 
> Even apart from this KEP, if clients are talking to the same DNS server expecting different DNS schemas, then that server will already need to know which version to serve up. So, I don't think we need to put it in the magic string - any pod using a different DNS policy would have this same problem, so we'll need to solve that if and when we change the schema version.
> 
outside of this KEP and without Autopath, the DNS Server does not need to know about versions. With the current autopath plugin implementation and CoreDNS, yes, if the searchpath schema changed, CoreDNS would need to know how to do this. But again, this becomes specific to Kubernetes. I think narrowing the scope to make this Kubernetes + CoreDNS specific might simplify some of the decisions here.

Not sure I understand how different DNSPolicy will have a problem. Apart from ClusterFirst*, the other modes do not use the kubelet-generated searchpaths. They are using the node's searchpath or something custom. Why would anything change there when the searchpath schema is changed?


> For the label, why not something like "ap.k8s.io" which is relatively short and we already control k8s.io so we can reserve that (and yes, it's three labels not one but I don't think that's an issue).
If we don't need version in the label, then ap.k8s.io is ok. Looks like there is a limit on the name length and label length, not number of labels.

Reply:
"schema version" here is not specific to this kep. For example, if you request TXT dns-version.cluster.local it will return a version number to you (well, CoreDNS does, i don't know that it was ever added to kube-dns, but it is in the spec).

There is an old issue (I can't find it though) discussing changing from names like `service.namespace.svc.cluster.local` to `service.svc.namespace.cluster.local`. That is what we mean by the schema version in this context.